master/internal/api_workspace.go

```diff
@@ -1317,6 +1318,94 @@ func getAutoGeneratedNamespaceName(ctx context.Context, wkspID int, tx *bun.Tx)
 	return &namespace, nil
 }
 
+func (a *apiServer) GetWorkspacesWithDefaultNamespaceBindings(ctx context.Context,
+	req *apiv1.GetWorkspacesWithDefaultNamespaceBindingsRequest,
+) (*apiv1.GetWorkspacesWithDefaultNamespaceBindingsResponse, error) {
+	k8ClusterNames := a.m.config.GetKubernetesClusterNames()
+	ns := db.Bun().NewSelect().
+		Table("workspace_namespace_bindings").
+		Column("workspace_id").
+		ColumnExpr("COUNT(*) as num_bindings").
+		Where("cluster_name IN (?)", bun.In(k8ClusterNames)).
+		GroupExpr("workspace_id")
+
+	var workspaceIds []int32
+	err := db.Bun().NewSelect().
+		With("ns", ns).
+		TableExpr("workspaces as w").
+		Column("w.id").
+		Join("LEFT JOIN ns on ns.workspace_id = w.id").
+		Where("ns.num_bindings < ? or ns.num_bindings IS NULL", len(k8ClusterNames)).
+		Scan(ctx, &workspaceIds)
+	if err != nil {
+		return nil, err
+	}
+
+	return &apiv1.GetWorkspacesWithDefaultNamespaceBindingsResponse{WorkspaceIds: workspaceIds}, nil
+}
+
+func (a *apiServer) BulkAutoCreateWorkspaceNamespaceBindings(ctx context.Context,
+	req *apiv1.BulkAutoCreateWorkspaceNamespaceBindingsRequest,
+) (*apiv1.BulkAutoCreateWorkspaceNamespaceBindingsResponse, error) {
+	k8ClusterNames := a.m.config.GetKubernetesClusterNames()
+	wsns := []model.WorkspaceNamespace{}
+	err := db.Bun().NewSelect().
+		Table("workspace_namespace_bindings").
+		Where("cluster_name IN (?)", bun.In(k8ClusterNames)).
+		Scan(ctx, &wsns)
+	if err != nil {
+		return nil, err
+	}
+	wkspBinding := map[int32][]string{}
```

Comment:
nit (non-blocking): It might be possible to handle the population of this struct in the SQL; however, by no means do you have to do it that way. It could be to be a trade off between potentially making the SQL query more complex and/or performance. 

If it's easier to approach this using go code, then let's keep this the way it is, but if you do something like an `ARRAY_AGG(cluster_name)` + a `GROUP BY (workspace_id)`, it might help avoid having to do the additional step of transforming the query result into the map.

Reply:
I think it is easier to just keep the go code for now, especially given the time crunch 😅

Reply:
Absolutely fair!